CI: CDash dashboard support

EZoni opened the pull request:
Trying to add and set up CDash dashboard support for WarpX. Close #5292.

Relevant documentation:
- https://cmake.org/cmake/help/book/mastering-cmake/chapter/CDash.html
- https://public.kitware.com/Wiki/CDash:Administration#Creating_a_project (project settings)
- https://cmake.org/cmake/help/latest/guide/tutorial/Adding%20Support%20for%20a%20Testing%20Dashboard.html

ax3l commented on CTestConfig.cmake:
```diff
@@ -0,0 +1,20 @@
+set(CTEST_PROJECT_NAME WarpX)
+set(CTEST_NIGHTLY_START_TIME 08:00:00 UTC)
+
+if(CMAKE_VERSION VERSION_GREATER 3.14)
+  set(CTEST_SUBMIT_URL https://my.cdash.org/submit.php?project=WarpX)
+else()
+  set(CTEST_DROP_METHOD "https")
+  set(CTEST_DROP_SITE "my.cdash.org")
+  set(CTEST_DROP_LOCATION "/submit.php?project=WarpX")
+endif()
```
```suggestion
set(CTEST_SUBMIT_URL https://my.cdash.org/submit.php?project=WarpX)
```

ax3l commented:
So far, just adding `-DExperimental` works to submit to CDash after CTest is done.

But... what if we want to get into the `Continuous` section instead of the `Experimental` one, can we do:
```
cmake -S . -B build  # as before

ctest --test-dir build -D ContinuousConfigure
ctest --test-dir build -D ContinuousBuild  # instead of cmake --build build?

ctest --test-dir build -D ContinuousTest  # instead of ctest --test-dir build
```

Still, does not submit and
```
ctest --test-dir build -D ContinuousSubmit
```
does not push it out...

EZoni commented:
I managed to get some better AI assistance and here's some useful information:

> Given that your Azure CI workflow runs `ctest` for each commit pushed to a PR and each PR merge to the main branch, you might want to use the `Experimental` dashboard type for the PRs and possibly `Continuous` or `Nightly` for the main branch merges.

Why using `Experimental` for PR events (e.g., push commit to an open PR)?
> `Experimental` Dashboard Type for PRs:
> 1. Purpose: The `Experimental` dashboard type is typically used for ad-hoc testing and development purposes. It is ideal for testing changes that are still under development and have not yet been merged into the main codebase.
> 2. Use Case: When a PR is created, it often contains changes that are still being reviewed and may undergo several iterations before being merged. Using the `Experimental` type for PRs allows you to track the results of these tests separately from the mainline tests. This helps in identifying and addressing issues early in the development process.
> 3. Benefits: It helps developers get immediate feedback on their changes without affecting the stability of the main branch's test results. It keeps the main dashboard clean and focused on the stability of the main branch, while still providing visibility into the ongoing development work.

Why using `Continuous` or `Nightly` for main branch events (e.g., merge PR into main branch)?
> `Continuous` or `Nightly` Dashboard Type for Main Branch Merges:
> 1. Purpose: The `Continuous` dashboard type is used for frequent, automated testing of the main codebase. It is suitable for running tests after each commit to the main branch to ensure that the codebase remains stable. The `Nightly` dashboard type is used for comprehensive testing that runs at a scheduled time each day. It is suitable for running more extensive tests that may take longer to complete.
> 2. Use Case: When changes are merged into the main branch, it is crucial to ensure that these changes do not introduce regressions or break existing functionality. Using the `Continuous` type for main branch merges allows for immediate testing and feedback on the stability of the main codebase. Using the `Nightly` type (if applicable) allows for more thorough testing, such as running a full suite of tests or performing additional checks that may not be feasible to run on every commit.
> 3. Benefits: It ensures that the main branch remains stable and reliable, as any issues introduced by recent changes can be quickly identified and addressed. It provides a clear and consistent view of the health of the main codebase over time.

EZoni commented on .azure-pipelines.yml:
```diff
@@ -155,8 +155,15 @@ jobs:
       set -o nounset errexit pipefail
-      # run tests (exclude pytest.AMReX when running Python tests)
-      ctest --test-dir build --output-on-failure -E AMReX
+      # determine if the build was triggered by a push to the development branch
+      if [[ "$(Build.SourceBranch)" == "refs/heads/development" ]]; then
+        # run tests (exclude pytest.AMReX when running Python tests)
+        # and submit results to CDash as Continuous
+        ctest --test-dir build --output-on-failure -E AMReX -D Continuous
+      else
+        # run tests (exclude pytest.AMReX when running Python tests)
+        ctest --test-dir build --output-on-failure -E AMReX
+      fi
```
This is what I would try first, based on the information reported in https://github.com/ECP-WarpX/WarpX/pull/5566#issuecomment-2596485383.

**Update**
This is probably duplicating the configure and build step. Looking more into how to avoid duplication.

EZoni replied:
Again on this, and on the issues reported in https://github.com/ECP-WarpX/WarpX/pull/5566#issuecomment-2594256505, reading the documentation more carefully it seems to me that the various "steps" (e.g., `Start`, `Update`, `Configure`, `Build`, `Test`, `MemoryCheck`, `Coverage`, `Submit`) can be run independently only for the `Nightly` or `Experimental` modes, but not for the other modes such as `Continuous`. This could explain what we observed so far.

![Screenshot from 2025-01-17 11-09-18](https://github.com/user-attachments/assets/3c2474c9-2fc8-415f-812c-cf783b7e2c89)

EZoni replied:
`Continuous` still seems like the right choice for main branch events, such as merging PRs into the main branch, from a theoretical point of view. But, as discussed, we could start with `Experimental` and fine tune things later on, once we have gained more experience with CDash overall.

EZoni replied:
```suggestion
      # determine if the build was triggered by a push to the development branch
      if [[ "$(Build.SourceBranch)" == "refs/heads/development" ]]; then
        # run tests (exclude pytest.AMReX when running Python tests)
        # and submit results to CDash as Experimental
        ctest --test-dir build --output-on-failure -E AMReX \
          -D ExperimentalTest -D ExperimentalSubmit
      else
        # run tests (exclude pytest.AMReX when running Python tests)
        ctest --test-dir build --output-on-failure -E AMReX
      fi
```

EZoni commented on CTestConfig.cmake:
```diff
@@ -0,0 +1,18 @@
+
+set(CTEST_DROP_SITE_CDASH TRUE)
+
+# Additional settings
+set(CTEST_SITE "Azure-Pipelines")
+set(CTEST_BUILD_NAME "CI-Development")
```
These site and build names would be displayed as follows:

![Screenshot from 2025-01-16 11-03-04](https://github.com/user-attachments/assets/21cdf381-f08d-456e-aaaa-2b5bec320288)

ax3l replied:
I wonder if these two should be set or if one can pass them to CMake/CTest... In the end, we will run the same file from Perlmutter and other sites, where the name and site does not really fit.

Or does one approach this with multiple config files (one per site)?

EZoni commented on .azure-pipelines.yml:
```diff
@@ -155,8 +155,16 @@ jobs:
-      # run tests (exclude pytest.AMReX when running Python tests)
-      ctest --test-dir build --output-on-failure -E AMReX
+      # determine if the build was triggered by a push to the development branch
+      if [[ "$(Build.SourceBranch)" == "refs/heads/development" ]]; then
```
If this does not work (but we'll know only after merging the PR into `development`), we can also try the following (see https://learn.microsoft.com/en-us/azure/devops/pipelines/build/variables for more documentation):
```suggestion
      if [[ "$(Build.Reason)" == "IndividualCI" ]]; then
```

ax3l replied:
You also need to check the repository name, to be sure (e.g., in case a fork develops on `development` and for some reason has Azure CI on).

ax3l commented on CMakeLists.txt:
```diff
@@ -193,7 +193,8 @@ if(WarpX_FFT)
 
-# this defined the variable BUILD_TESTING which is ON by default
+# Define the variable BUILD_TESTING (ON by default),
+# include CDash dashboard testing module
```
```suggestion
# include CDash dashboard testing module (CTestConfig.cmake)
```

ax3l reviewed (approved):
I'll merge this for now for testing, let's address the inline comments in follow-ups.
